src/main/java/org/myrobotlab/service/Mqtt.java

```diff
@@ -14,231 +18,390 @@
 
-        if (userName != null) {
-          conOpt.setUserName(userName);
-          conOpt.setPassword(password);
-        }
+      if (username != null && username.length() > 0) {
+        conOpt.setUserName(username);
+        conOpt.setPassword(password.toCharArray());
+      }
 
-        client.setCallback(this);
-        client.connect(conOpt, userContext, this);
-        int i = 0;
+      client.setCallback(this);
 
-        // wait for a CONNACK !
-        while (!client.isConnected() && i < 10) {
-          sleep(100);
-          i += 1;
-        }
+      // onOpt.setMqttVersion(MqttConnectOptions.MQTT_VERSION_3_1);
+      // conOpt.setMqttVersion(MqttConnectOptions.MQTT_VERSION_3_1_1);
 
-        if (client.isConnected()) {
-          isConnected = true;
-        } else {
-          isConnected = false;
-          if (!autoReconnect) {
-            broadcastState();
-            return;
-          }
-        }
+      if (sslCertFilePath != null && sslKeyFilePath != null) {
+        SSLSocketFactory socketFactory = SslUtil.getSocketFactory(sslCaFilePath, sslCertFilePath, sslKeyFilePath, sslPassword);
+        conOpt.setSocketFactory(socketFactory);
+      }
+
+      // default is 10 - which drops messages on a fast local network :(
+      conOpt.setMaxInflight(1000);
+
+      client.connect(conOpt, userContext, this);
+
+      int i = 0;
+
+      // wait for a CONNACK !
+      while (!client.isConnected() && i < 10 /* (autoReconnect || i < 10) */) {
+        sleep(500);
+        i += 1;
+      }
 
-        // successful connection
+      // DO THIS ON CONNECTED EVENT !!!- does it exist
+      if (client.isConnected()) {
+        connected = true;
+        info("connected to %s", url);
 
         // subscribe now
-        if (autoSubscribe) {
-          subscribe(inTopic + "/#");
-          info("subscribed to topic %s", inTopic);
+        if (autoConnect) {
+          // announce to all we have connected
+          broadcastConnect();
+          // broadcast every 10 seconds
+          if (broadcastConnectPingIntervalMs != null) {
+            addTask(broadcastConnectPingIntervalMs, "broadcastConnect");
+          }
+
+          // subscribe to the announce connect topic
+          subscribe(onConnectTopic); // TODO - use LWT topics !
+          subscribe(onDisconnectTopic);
+
+          // subscribe to "general" rx topic
+          String rxTopic = String.format("mrl/gw/%s/rx", getId());
+          info("creating general rx topic %s", rxTopic);
+          subscribe(rxTopic); // TODO - use LWT topics !
         }
+
         broadcastState();
+        return;
+      } else {
 
-      } catch (Exception e) {
-        sleep(1500);
-        log.error("connect failed", e);
+        error("could not connect to %s", url);
+        connected = false;
+        if (!autoReconnect) {
+          broadcastState();
+          return;
+        }
       }
+
+      broadcastState();
+
+    } catch (Exception e) {
+      Service.sleep(1500);

```

Comment:
why sleep here at all? the exception is already blown.

Reply:
heh ... I don't know how many times I've seen reconnect logic without a sleep, and the consequence usually is a fast spin that pegs the cpu when the host is unreachable, rather than retrying at a meaningful interval (or progressive backoff) the client pegs the cpu the whole process or whole computer become bork'd...

But this is an artifact from my first implementation, I later discovered the paho client itself implemented reconnect logic - so ya ... mebbe can be removed ... I look

Reply:
I removed it .. I'm going to rely on the paho reconnect and see how that goes